master/internal/api_experiment.go

```diff
@@ -1399,6 +1399,272 @@ func (a *apiServer) createUnmanagedExperimentTx(
 	}, nil
 }
 
+func (a *apiServer) parseAndMergeContinueConfig(expID int, overrideConfig string) ([]byte, error) {
+	if overrideConfig == "" {
+		overrideConfig = "{}"
+	}
+
+	providedConfig, err := expconf.ParseAnyExperimentConfigYAML([]byte(overrideConfig))
+	if err != nil {
+		// Add a helpful error message if a user just submits
+		// searcher.max_length.batches = 2. They would also need
+		// searcher.name = "single", which all experiments will always be here.
+		if strings.Contains(err.Error(), `unknown field "max_length"`) {
+			return nil, status.Errorf(codes.InvalidArgument,
+				`unknown field "max_length", you might also need to specify searcher.name=single`)
+		}
+
+		return nil, status.Errorf(codes.InvalidArgument,
+			fmt.Errorf("parsing override config: %w", err).Error())
+	}
+
+	if providedConfig.RawProject != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "'project' in override config "+
+			"cannot be specified, use `det experiment move` first if you want to change the project")
+	}
+	if providedConfig.RawWorkspace != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "'workspace' in override config "+
+			"cannot be specified, use `det experiment move` first if you want to change the workspace")
+	}
+	// TODO deny list more stuff.
+
+	activeConfig, err := a.m.db.ActiveExperimentConfig(expID)
+	if err != nil {
+		return nil, fmt.Errorf("loading active config for experiment %d: %w", expID, err)
+	}
+	if name := activeConfig.Searcher().AsLegacy().Name; name != "single" { //nolint: goconst
+		return nil, status.Errorf(codes.InvalidArgument,
+			fmt.Sprintf(
+				"cannot continue a '%s' searcher experiment, must be a single searcher experiment",
+				name))
+	}
+
+	mergedConfig := schemas.Merge(providedConfig, activeConfig)
+	if name := mergedConfig.Searcher().AsLegacy().Name; name != "single" { //nolint: goconst
+		return nil, status.Errorf(codes.InvalidArgument,
+			fmt.Sprintf("override config must have single searcher type got '%s' instead", name))
+	}
+
+	bytes, err := mergedConfig.Value()
+	if err != nil {
+		return nil, fmt.Errorf("getting value of merged config: %w", err)
+	}
+
+	return bytes.([]byte), nil
+}
+
+func (a *apiServer) errorIfSearcherIsDone(trialID int, e expconf.ExperimentConfig) error {
+	latestCheckpoint, err := a.m.db.LatestCheckpointForTrial(trialID)
+	if err != nil {
+		return fmt.Errorf("getting latest checkpoint for trialID %d: %w", trialID, err)
+	}
+	if latestCheckpoint == nil || latestCheckpoint.StepsCompleted == 0 {
+		return nil // Can't be done if we don't have a checkpoint.
+	}
+
+	if e.Searcher().RawSingleConfig == nil {
+		return fmt.Errorf("something went wrong searcher single config is nil")
+	}
+
+	var inBatches int
+	units := int(e.Searcher().RawSingleConfig.MaxLength().Units)
+	unit := e.Searcher().Unit()
+	switch unit {
+	case expconf.Unitless, expconf.Unspecified:
```

Comment:
Is this reasonable to just let unitless through?

For example if you try to resume `tutorials/core_api/2_checkpoints.yaml`

```
[2023-08-30T15:26:27.184751Z] b8716e83 || Traceback (most recent call last):
[2023-08-30T15:26:27.184758Z] b8716e83 ||   File "2_checkpoints.py", line 90, in <module>
[2023-08-30T15:26:27.186659Z] b8716e83 ||     main(
[2023-08-30T15:26:27.186661Z] b8716e83 ||   File "2_checkpoints.py", line 75, in main
[2023-08-30T15:26:27.186712Z] b8716e83 ||     core_context.train.report_validation_metrics(steps_completed=steps_completed, metrics={"x": x})
[2023-08-30T15:26:27.186713Z] b8716e83 || UnboundLocalError: local variable 'steps_completed' referenced before assignment
[2023-08-30T15:26:27.841264Z] b8716e83 || ERROR: crashed: resources failed with non-zero exit code: container failed with non-zero exit code: 1 (exit code 1)
[2023-08-30T15:26:27.844294Z]          || ERROR: Trial 44 (Experiment 44) was terminated: allocation failed: resources failed with non-zero exit code: container failed with non-zero exit code: 1 (exit code 1)
```

It fails because it doesn't expect to have a starting batch equal to the ending batch. Though I think if something goes wrong with core API that is just the users responsibility to account for it if they want to use the feature?